pkg/promtail/targets/filetarget.go

```diff
@@ -168,164 +142,161 @@ func (t *FileTarget) run() {
+func (t *FileTarget) stopWatching(dirs map[string]struct{}) {
+	for dir := range dirs {
+		if _, ok := t.watches[dir]; !ok {
+			continue
+		}
+		level.Debug(t.logger).Log("msg", "removing directory from watcher", "directory", dir)
+		err := t.watcher.Remove(dir)
+		if err != nil {
+			level.Error(t.logger).Log("msg", " failed to remove directory from watcher", "error", err)
+		}
+		// Shutdown and cleanup and tailers for files in directories no longer being watched.
+		for tailedFile := range t.tails {
+			if filepath.Dir(tailedFile) == dir {
+				t.stopTailing(tailedFile)
 			}
+		}
+	}
+}
 
-			readLines.WithLabelValues(t.path).Inc()
-			readBytes.WithLabelValues(t.path).Add(float64(len(line.Text)))
-			if err := t.handler.Handle(model.LabelSet{}, line.Time, line.Text); err != nil {
-				level.Error(t.logger).Log("msg", "error handling line", "path", t.path, "error", err)
-			}
-		case <-t.quit:
-			return
+func (t *FileTarget) updateTailers(matches []string) {
+	// Start tailing all of the matched files if not already doing so.
+	for _, p := range matches {
+		err := t.startTailing(p)
+		if err != nil {
+			level.Error(t.logger).Log("msg", "sync() failed to start tailer", "error", err, "filename", p)
+			continue
 		}
 	}
+
+	// Stop tailing any files which no longer exist
+	existingTails := map[string]struct{}{}
+	for file := range t.tails {
+		existingTails[file] = struct{}{}
+	}
+	currentMatches := map[string]struct{}{}
+	for _, file := range matches {
+		currentMatches[file] = struct{}{}
+	}
+	toStopTailing := missing(currentMatches, existingTails)
+	for p := range toStopTailing {
+		t.stopTailing(p)
+	}
 }
 
-func (t *tailer) markPosition() error {
-	pos, err := t.tail.Tell()
+func (t *FileTarget) startTailing(path string) error {
+	if _, ok := t.tails[path]; ok {
+		return nil
+	}
+	fi, err := os.Stat(path)
+	if err != nil {
+		return errors.Wrap(err, "filetarget.startTailing.stat")
+	}
+	if fi.IsDir() {
+		return errors.New("File is a directory and cannot be tailed")
+	}
+	level.Debug(t.logger).Log("msg", "tailing new file", "filename", path)
+	tailer, err := newTailer(t.logger, t.handler, t.positions, path)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "filetarget.startTailing.newTailer")
 	}
-	level.Debug(t.logger).Log("path", t.path, "filename", t.filename, "current_position", pos)
-	t.positions.Put(t.filename, pos)
+	t.tails[path] = tailer
 	return nil
 }
 
-func (t *tailer) stop() error {
-	close(t.quit)
-	<-t.done
-	filesActive.Add(-1.)
-	return t.tail.Stop()
+func (t *FileTarget) stopTailing(path string) {
+	tailer, ok := t.tails[path]
+	if ok {
+		helpers.LogError("stopping tailer", tailer.stop)
+		tailer.cleanup()
+		delete(t.tails, path)
+	}
 }
 
-func (t *tailer) cleanup() {
-	t.positions.Remove(t.filename)
+// Returns the elements from set b which are missing from set a
+func missing(a map[string]struct{}, b map[string]struct{}) map[string]struct{} {
+	c := map[string]struct{}{}
+	for dir := range b {
```

Comment:
Super nit: `dir` name is a bit weird.  I tend to go with `as`, `bs` and then call them `a` and `b` in the loop.